### PR TITLE
Use AnyIO

### PR DESCRIPTION
**Pull request by davidbrochart**
This PR has breaking changes, and should probably be released as v1.0.
In particular, `WebsocketProvider`, `WebsocketServer`, `YRoom` and `YStore` must either be used with an async context manager (this is the preferred way), or using lower-level `start()` and `stop()` methods.
Before this PR, some tasks were created implicitly on instanciation. The new approach is more explicit regarding the async nature of these objects, and ensures no task is left running on tear-down. Under the hood, [AnyIO](https://anyio.readthedocs.io)'s task groups are used, but ypy-websocket can be used in a "pure asyncio" environment, no need to adopt AnyIO outside of this library.
Here is an example with an async context manager:
```py
async def main():
    async with WebsocketServer():
        ...
```
Which is equivalent to the following, when using the lower-level API (with `start()`/`stop()`):
```py
async def main():
    server = WebsocketServer()
    task = asyncio.create_task(server.start())
    await server.started.wait()
    ...
    server.stop()
```

Closes #76.

### CODE REVIEW

File: ypy_websocket/websocket_server.py [hunk 1]
```diff
@@ -100,62 +32,92 @@ def get_room_name(self, room):
...
-        # remove this client
-        room.clients = [c for c in room.clients if c != websocket]
-        if self.auto_clean_rooms and not room.clients:
-            self.delete_room(room=room)
+        if self._task_group is None:
+            raise RuntimeError(
+                "The WebsocketServer is not running: use `async with websocket_server:` or `await websocket_server.run()`"
+            )
+
+        await self._task_group.start(self._serve, websocket)
```
Comment (fcollonval) [inline ypy_websocket/websocket_server.py:58]: Why not starting to serve directly? The need to do

```py
await websocket_server.run()
await websocket_server.serve()
``` 

seems error prone to me.

File: ypy_websocket/ystore.py [hunk 2]
```diff
@@ -35,6 +35,23 @@ async def write(self, data: bytes) -> None:
...
+            tg = anyio.create_task_group()
+            self._task_group = await exit_stack.enter_async_context(tg)
+            self._exit_stack = exit_stack.pop_all()
+            tg.start_soon(self.start)
+
+    async def __aexit__(self, exc_type, exc_value, exc_tb):
+        self._task_group.cancel_scope.cancel()
+        return await self._exit_stack.__aexit__(exc_type, exc_value, exc_tb)
+
+    async def start(self):
```
Comment (fcollonval) [inline ypy_websocket/ystore.py:49]: Same question about naming alignment and protection of the two API usage.

**Review by fcollonval — COMMENTED** [review]
Thanks @davidbrochart 

I have questions trying to understand all changes. It would also be great to add more documentation to the code especially as it is meant to be extended by other libraries.

File: ypy_websocket/websocket_server.py [hunk 3]
```diff
@@ -100,62 +32,92 @@ def get_room_name(self, room):
...
+                        )
+                        for client in room.clients:
+                            self.log.debug(
+                                "Sending Y awareness from client with endpoint %s to client with endpoint: %s",
+                                websocket.path,
+                                client.path,
+                            )
+                            tg.start_soon(client.send, message)
+            except Exception:
+                pass
```
[the remark this answers is not in the document]
Comment (davidbrochart) [reply ypy_websocket/websocket_server.py:108]: :+1: will do

File: ypy_websocket/websocket_provider.py [hunk 4]
```diff
@@ -16,29 +19,49 @@
...
+                    await self._websocket.send(message)
+                except Exception:
+                    pass
+
+    async def run(self):
+        async with create_task_group() as self._task_group:
+            self._task_group.start_soon(self._run)
+
+    def stop(self):
+        self._task_group.cancel_scope.cancel()
```
[the remark this answers is not in the document]
Comment (davidbrochart) [reply ypy_websocket/websocket_provider.py]: I want to allow both an async context manager API (that should be the preferred way) and a lower-level API with `start()` and `stop()` equivalent methods. It is a common practice. I'll update the PR description.

File: ypy_websocket/websocket_server.py (hunk 1 above)
Comment (davidbrochart) [reply ypy_websocket/websocket_server.py]: The `serve()` method serves one WebSocket, but the server can serve multiple WebSockets. The goal of the (now called) `start()` method is to create the main task group, whose cancel scope is cancelled when `stop()` is called or when the async context manager exits. This ensures no task is running when the `WebsocketServer` stops.

File: ypy_websocket/yroom.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    async def enter(self):
+        if self._entered:
+            return
+
+        async with create_task_group() as self._task_group:
+            self._task_group.start_soon(self._broadcast_updates)
+            self._entered = True
+
+    def exit(self):
+        self._task_group.cancel_scope.cancel()
```
[the remark this answers is not in the document]
Comment (davidbrochart) [reply ypy_websocket/yroom.py]: > Would it make sense to align the wording using `run`/`stop`?

Yes, it's done in https://github.com/y-crdt/ypy-websocket/pull/77/commits/f0fa069908c9dfbb1f9782abbbcda809b947227c.

> It seems you implemented a flag to ensure `enter` is called only once, why not implementing the flag completely and prevent exit if not enter was called or avoiding clashing between using the object in async context manager and public API -- this applies to the other async context manager objects too.

Good point, I'll make the code more robust in that regard, thanks.

File: ypy_websocket/ystore.py (hunk 2 above)
Comment (davidbrochart) [reply ypy_websocket/ystore.py:65]: :+1:

**Comment by davidbrochart** [comment]
> I have questions trying to understand all changes. It would also be great to add more documentation to the code especially as it is meant to be extended by other libraries.

Thanks for the review @fcollonval,  I updated the PR description. Where do you think I should add comments in the code in particular?

**Comment by davidbrochart** [comment]
The documentation is accessible at https://davidbrochart.github.io/ypy-websocket.

**Comment by fcollonval** [comment]
> best would be to use mermaidJS so the graph are nicely display directly in GitHub.

**Pull request merged**